cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -83,29 +90,64 @@ public PushAppResponse push(PushAppOperationRequest pushRequest) throws NoChange
             pushRequest.getNamespaceId(),
             pushRequest.getRepositoryConfig())
     ) {
-      try {
-        repositoryManager.cloneRemote();
-      } catch (GitAPIException | IOException e) {
-        throw new GitOperationException(String.format("Failed to clone remote repository: %s",
-                                                       e.getMessage()), e);
-      }
+      cloneAndCreateBaseDir(repositoryManager);
+      ApplicationDetail app = pushRequest.getApp();
+      writeAppDetail(repositoryManager, app);
 
-      LOG.info("Pushing application configs for : {}", pushRequest.getApp().getName());
+      List<PushAppResponse> responses = new ArrayList<>(commitAndPush(repositoryManager,
+          pushRequest.getCommitDetails(),
+          ImmutableList.of(new PushFile(
+                  app.getName(),
+                  app.getAppVersion(),
+                  repositoryManager.getFileRelativePath(generateConfigFileName(app.getName()))
+              )
+          )
+      ));
 
-      //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
-      return writeAppDetailAndPush(
-          repositoryManager,
-          pushRequest.getApp(),
-          pushRequest.getCommitDetails()
-      );
+      // it should never be empty as in case of any error we will get an exception
+      return responses.get(0);
+    }
+  }
+
+  @Override
+  public Collection<PushAppResponse> multiPush(MultiPushAppOperationRequest pushRequest,
+      ApplicationManager appManager)
+      throws NoChangesToPushException, AuthenticationConfigException {
+    try (
+        RepositoryManager repositoryManager = repoManagerFactory.create(
+            pushRequest.getNamespaceId(),
+            pushRequest.getRepositoryConfig())
+    ) {
+      cloneAndCreateBaseDir(repositoryManager);
+
+      LOG.info("Pushing application configs for {} apps.", pushRequest.getApps().size());
+
+      List<PushFile> filesToPush = new LinkedList<>();
+
+      for (String appToPush : pushRequest.getApps()) {
+        LOG.info("Pushing application configs for : {}.", appToPush);
+        ApplicationReference appRef = new ApplicationReference(pushRequest.getNamespaceId(),
+            appToPush);
+        try {
+          ApplicationDetail detail = appManager.get(appRef);
```

Comment:
This loop would be very slow? It is getting one app at a time from the worker remotely? We should add a batch method to the manager.

Reply:
Currently this operation uses only the `LocalApplicationManager`, which gets the `ApplicationDetail` in memory. The `RemoteApplicationManager` will be used here, only when the corresponding LongRunningOperationRunner is implemented. @samdgupi can provide more information on it.

Reply:
That’s why the implementation here is not safe. We should have a batch fetch method

Reply:
AFAIK We currently have no API/ which can fetch details by list of ids. We had plan to do this optimization after initial release of the feature. I will add this a TODO and take it up if possible as part of current release.